existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,194 @@
+This is useful not only in such big examples, but also when your function requires any of its arguments to be specified on usage.
+
+#### Additional type variables
+Second, you now can define functions with more type variables than you use on the right side of declaration.
+This is used mostly when you want to add complex constraint, which require several types, without adding them into the definition itself:
+
+``` Haskell
+someMonadicFunction :: forall a m env. MonadReader env m => a -> m a
+```
+
+Here we don't have `env` used in the type definition, but despite that, we are still able to add `MonadReader` constraint.
+
+## Existential quantification
+Beside universally quantified variables, Haskell also supports existential types.
+In order to use them ... you still need to use `forall.` keyword ¯\\\_(ツ)_/¯.
+That might be a little confusing.
+The following construction is called existential type in Haskell - when you use a forall quantifier in the right-hand side of function or data type declaration.
+But what is it used for? Let's find out!
+
+### Usage
+
+We will start with functions:
+
+```haskell
+func :: forall a c. a -> (forall b. (Show b) => b) -> c
+```
+Few things to mention.
+Here `a` and `c` are universally quantified and `b` is existentially quantified variables.
+The same variable can't be quantified more than once.
+Also, you can see that constraints for `b` are added in the middle of the function type declaration.
+This example is pretty useless.
+The only thing that has changed here is that now you can't manually choose type for `b` - the compiler must deduce the correct variable you want to use by itself.
+
+When we come to data type declaration, it becomes more useful.
+
+#### Hiding type variables
+```Haskell
+data MyData a = forall b. MkMyData b a
+```
+The first thing you might have noticed is that we don't have a type variable for `b` on the left side of expression.
+However, we still can use it on the right side.
+This is because `b` is existentially quantified.
+
+```Haskell
+MkMyData @_ @Integer -- This won't work, you don't have second variable in declaration
+```
+Compiler will deduce the type of the variable itself, and you can't specify it.
+
+That is actually why this construction is called existential type - because you tell a compiler, that there exists type, that
+can be used in constructor, without specifying which one exactly. Also it is called so, because from type theory's point of view
+this constructions are isomorphic:
+
+```haskell
+data MyTypeWithFunc = forall a. MkMyTypeWithFunc a (a -> String)
+
+MkMyTypeWithFunc :: forall a . a -> (a -> String)-> MyTypeWithFunc
+MkMyTypeWithFunc :: (exists a . (a, a -> String)) -> MyTypeWithFunc
+```
+
+#### Heterogeneous lists
+
+Now we understand that existential types allow you to introduce variables, without necessity to specify them.
+This might be useful when building a multi-typed list:
```

Comment:
_Why_ would the user want to do this? The `Elem` example is not very motivating, there's literally nothing you can do with a value of type `Elem`. So having a list of `[Elem]` is also useless.

Perhaps a more motivating example would be to create a list of things that can be displayed on the screen:

```hs
data Showable = forall a. Show a => Showable a

printMany :: [Showable] -> IO ()
printMany xs = 
  forM_ xs $ \(Showable x) -> print x
```

```
λ> printMany [Showable 1, Showable "hello"]
1
"hello"
```